### ISSUE TITLE
exiv2 0.28.0 support

### ISSUE THREAD

**Issue by 0-wiz-0** [BestImageViewer/geeqie#1090]
### ISSUE TYPE

* Bug Report

### GEEQIE VERSION

```
2.0.1
```

```
geeqie 2.0.1

  Documentation
    help                 : yelp-build not found - Help files created:
                           NO
    nl_langinfo          : nl_langinfo not found - first weekday depends on locale:
                           NO
                           first weekday defaults to Monday
    ChangeLog            : ChangeLog, ChangeLog.html created:
                           NO
    README               : pandoc not found - README.html created:
                           NO
    print preview        : evince not found - print preview supported:
                           NO

  Directories
    gq_appdir            : share/geeqie
    gq_bindir            : share/doc/geeqie
    gq_helpdir           : share/doc/geeqie
    gq_htmldir           : share/doc/geeqie/html
    gq_localedir         : share/locale

  Configuration
    archive              : libarchive >=3.4.0 not found - archive files e.g. .zip supported::
                           NO
    cms                  : color management supported:
                           YES
    djvu                 : ddjvuapi >=2.5.27 not found - djvu files supported:
                           NO
    exiv2                : image metadata processed by exiv2:
                           YES
    videothumbnailer     : libvideothumbnailer >=2.1.0 not found - thumbnails of video files supported
                           NO
    gps-map-champlain    : champlain-0.12 >=0.12 not found - GPS map displayed:
                           NO
    gps-map-champlain-gtk: champlain-gtk-0.12 >=0.12 not found - GPS map displayed:
                           NO
    heif                 : heif files supported:
                           YES
    j2k                  : j2k files supported:
                           YES
    jpeg                 : jpeg files supported:
                           YES
    jpegxl               : libjxl >=0.3.7 not found - jpegxl files supported:
                           NO
    libraw               : libraw >=0.20 not found - .cr3 files supported:
                           NO
    lua                  : lua5.3 >=5.3 not found - lua supported:
                           NO
    pdf                  : poppler-glib >=0.62 not found - pdf files supported:
                           NO
    spell                : gspell-1 >=1.6 not found - spelling checks not enabled
                           NO
    tiff                 : libtiff not found - tiff files supported:
                           NO
    webp                 : webp files supported:
                           YES

  User defined options
    buildtype            : plain
    libdir               : lib
    libexecdir           : libexec
    mandir               : man
    prefix               : /usr/pkg
    sysconfdir           : /usr/pkg/etc
    doxygen              : disabled

```

### OS / DISTRIBUTION

NetBSD/10.99.4/amd64 using pkgsrc

### SUMMARY

exiv2 0.28.0 changed some interfaces which broke geeqie's exiv2 code.

### STEPS TO REPRODUCE

Try building with exiv2 0.28.0
[geeqie.log](https://github.com/BestImageViewer/geeqie/files/11496088/geeqie.log)

I've asked upstream for a portability guide for 0.28.0 here:
https://github.com/Exiv2/exiv2/issues/2630

**Comment by neheb** [comment]
```
--- a/src/exiv2.cc
+++ b/src/exiv2.cc
@@ -62,6 +62,11 @@

 #include "misc.h"

+#if EXIV2_TEST_VERSION(0,28,0)
+#define AnyError Error
+#define AutoPtr UniquePtr
+#endif
+
 typedef struct _AltKey AltKey;

 struct _AltKey
@@ -174,7 +179,7 @@ public:
        {
                cp_data_ = NULL;
                cp_length_ = 0;
-               image_ = image;
+               image_ = std::move(image);
                valid_ = TRUE;
        }

@@ -364,7 +369,11 @@ public:
                        Exiv2::Image *image = imageData_->image();

 #ifdef HAVE_EXIV2_ERROR_CODE
+#if EXIV2_TEST_VERSION(0,28,0)
+                       if (!image) throw Exiv2::Error(Exiv2::ErrorCode::kerInputDataReadFailed);
+#else
                        if (!image) throw Exiv2::Error(Exiv2::kerInputDataReadFailed);
+#endif
 #else
                        if (!image) throw Exiv2::Error(21);
 #endif
@@ -839,7 +848,11 @@ gint exif_item_get_integer(ExifItem *item, gint *value)
 {
        try {
                if (!item || exif_item_get_elements(item) == 0) return 0;
+#if EXIV2_TEST_VERSION(0,28,0)
+               *value = ((Exiv2::Metadatum *)item)->toInt64();
+#else
                *value = ((Exiv2::Metadatum *)item)->toLong();
+#endif
                return 1;
        }
        catch (Exiv2::AnyError& e) {
@@ -1224,10 +1237,17 @@ guchar *exif_get_preview(ExifData *exif, guint *data_len, gint requested_width,
                        Exiv2::PreviewImage image = pm.getPreviewImage(*pos);

                        Exiv2::DataBuf buf = image.copy();
+#if EXIV2_TEST_VERSION(0,28,0)
+                       *data_len = buf.size();
+                       auto b = buf.data();
+                       buf.reset();
+                       return b;
+#else
                        std::pair<Exiv2::byte*, long> p = buf.release();

                        *data_len = p.second;
                        return p.first;
+#endif
                        }
                return NULL;
        }
```

**Comment by xsdg** [comment]
I haven't taken a close look, but "initializer but incomplete type" means that the type is being used in a context where it needs to be fully defined, but the compiler doesn't see a full definition yet at that point.

This is a pure guess, but I'm guessing that `error.hpp` relies on a type that's defined in `log.hpp` or something, but doesn't actually `#include` that dependency, and it probably wasn't noticed because everywhere else that uses it also includes `log.hpp` first.

**Comment by xsdg** [comment]
(That is, just by the signature, this looks like an exiv2 bug)

**Comment by neheb** [comment]
Weird.

```
mkdir subprojects
meson wrap install exiv2
meson setup --wrap_mode=forcefallback build
meson compile -C build
```

Should result in a successful build.

**Comment by caclark** [comment]
Thanks @neheb . I can now compile the project. But....

With the exif window open and viewing a certain type of file there is a crash. It happens with NEF files from my camera, but not with NEF test images downloaded from the Internet.

Attached is the log file and a sample image. If anyone can easily see the problem, that is good. Otherwise it will take me a while (or longer) to find the problem.

[log3.txt.gz](https://github.com/BestImageViewer/geeqie/files/11581664/log3.txt.gz)
[2019-12-23-11:21:44-DSC_3328.NEF.gz](https://github.com/BestImageViewer/geeqie/files/11581666/2019-12-23-11.21.44-DSC_3328.NEF.gz)

**Comment by neheb** [comment]
Sounds like an upstream issue

**Comment by caclark** [comment]
Compilation is ok, but the seg. fault with some .nef files remains.

**Comment by caclark** [comment]
I have raised an issue at exiv2: https://github.com/Exiv2/exiv2/issues/2638

If I knew anything about c++, things would be easier....

**Comment by neheb** [comment]
since this project is C++ as well, probably should use clang-tidy.

**Comment by caclark** [comment]
Fixed in commit c45cca77
